inst/include/dust/dust.hpp

```diff
@@ -145,22 +135,31 @@ class Dust {
     for (size_t i = 0; i < n_particles; ++i) {
       particles_[i].set_step(step);
     }
+    stale_device_ = true;
   }
 
-  void set_step(const std::vector<size_t>& step, const bool deterministic) {
+  void set_step(const std::vector<size_t>& step) {
     const size_t n_particles = particles_.size();
     for (size_t i = 0; i < n_particles; ++i) {
       particles_[i].set_step(step[i]);
     }
     const auto r = std::minmax_element(step.begin(), step.end());
     if (*r.second > *r.first) {
-      run(*r.second, deterministic);
+      run(*r.second);
     }
+    stale_device_ = true;
   }
 
-  void run(const size_t step_end, const bool deterministic) {
+  // It's the callee's responsibility to ensure that index is in
+  // range [0, n-1]
+  void set_index(const std::vector<size_t>& index) {
+    index_ = index;
+    update_device_index();
+  }
+
+  void run(const size_t step_end) {
     refresh_host();
-    if (deterministic) {
+    if (deterministic_) {
```

Comment:
Now this is no longer dynamic, should `rng_.set_deterministic()` be moved to the constructor?

Reply:
You can't tell on the constructor if the model is going to be used on the device (we accept `NULL` device information)

Reply:
oh sorry, it is there. I must be doing something daft here

Reply:
Right; we need the rng to be usable in a non-deterministic way for compare and theoretically for the filter, so this dynamism is required here, even though it's a bit surprising. The consistent way of doing this would be I think to check at set_data and prevent anyone adding data, which in turn prevents use of compare/filter